### PR TITLE
SMD-718: Add validation for Unit of Measurement

### PR DESCRIPTION
**Pull request by MarcUsher**
### Change description
- Add cross-table validation for output/outcome unit of measurement
- UoM validation only happens on outputs/outcomes that themselves don't have a validation error
- Update error message templates to match updated copy from design team
- Update tests to reflect new error messages and check new UoM validation
- Update test mock configs and mock_df_dict for pathfinders as data was inconsistent with final spreadsheet

- [X] Unit tests and other appropriate tests added or updated
- [ ] README and other documentation has been updated / added (if needed)
- [X] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
_If manual testing is needed, give suggested testing steps_

### CODE REVIEW

**Review by Joel-Pearce — CHANGES_REQUESTED** [review]

File: core/validation/specific_validation/pathfinders/round_1.py [hunk 1]
```diff
@@ -163,23 +163,52 @@ def _check_standard_outputs(
...
         allowed_values_map=control_mappings["intervention_theme_to_standard_outputs"],
     )
-    breaching_outputs = extracted_table_dfs["Outputs"].loc[breaching_row_indices, "Output"].tolist()
-    return [
+    breaching_outputs = (
+        extracted_table_dfs["Outputs"]
+        .loc[breaching_row_indices_outputs, ["Output", "Intervention theme"]]
+        .values.tolist()
+    )
+    breaching_indices_copy = breaching_row_indices_outputs.copy()
```
[the remark this answers is not in the document]
Comment (MarcUsher) [reply core/validation/specific_validation/pathfinders/round_1.py:177]: V good point, I have tested it but it should be `deepcopy()` for safety and I'll give it some more testing

File: core/validation/specific_validation/pathfinders/round_1.py [hunk 2]
```diff
@@ -163,23 +163,52 @@ def _check_standard_outputs(
...
+        allowed_values_key_column="Output",
+        allowed_values_map=control_mappings["standard_output_uoms"],
+    )
+    breaching_uoms = extracted_table_dfs["Outputs"].loc[breaching_row_indices_uom, "Unit of measurement"].tolist()
+    uom_errors = [
+        _error_message(
+            sheet="Outputs",
+            section="Standard outputs",
+            description=PFErrors.UOM_NOT_ALLOWED.format(unit_of_measurement=unit_of_measurement),
+            cell_index=f"D{breaching_row_indices_uom.pop(0) + 1}",
```
[the remark this answers is not in the document]
Comment (MarcUsher) [reply core/validation/specific_validation/pathfinders/round_1.py:205]: Agree we should mention it wherever it crops in, will update to include in the docstring

**Review by Joel-Pearce — CHANGES_REQUESTED** [review]
The docstrings need updating where UoM is now something that is also being looked at in tandem with intervention theme (they only mention the latter)

File: core/transformation/pathfinders/round_1/control_mappings.py [hunk 3]
```diff
@@ -104,3 +108,43 @@ def _intervention_theme_to_standard_outcomes(standard_outcomes_df: pd.DataFrame)
...
+    }
+
+
+def _bespoke_outcome_uoms(bespoke_outcomes_df: pd.DataFrame) -> dict[str, str]:
+    """Creates a mapping from bespoke outcome to its unit of measurement."""
+    return {
+        row["Outcome"]: bespoke_outcomes_df.loc[bespoke_outcomes_df["Outcome"] == row["Outcome"], "UoM"].tolist()
+        for _, row in bespoke_outcomes_df.iterrows()
+        if not pd.isna(row["UoM"])
+    }
```
Comment (Joel-Pearce) [inline core/transformation/pathfinders/round_1/control_mappings.py:150]: Could we not just combine these and similar into one function calling params on the name of the columns?

Or do you reckon it's better to be verbose like this?

**Review by Joel-Pearce — CHANGES_REQUESTED** [review]

File: core/transformation/pathfinders/round_1/control_mappings.py [hunk 4]
```diff
@@ -104,3 +108,17 @@ def _intervention_theme_to_standard_outcomes(standard_outcomes_df: pd.DataFrame)
...
+
+    :param control_data_df: Dataframe of the extracted control data table
+    :param column_name: String value of the column name from the relevant control table to be used for the mapping
+    :return: Dictionary of output or outcome names to a list of their unit of measurement
+    """
+    return {
+        row[column_name]: control_data_df.loc[control_data_df[column_name] == row[column_name], "UoM"].tolist()
+        for _, row in control_data_df.iterrows()
+        if not pd.isna(row["UoM"])
+    }
```
Comment (Joel-Pearce) [inline core/transformation/pathfinders/round_1/control_mappings.py:124]: I can't see a unit test for this function, would it be possible to create one? Or am I missing something?

**Review by Joel-Pearce — CHANGES_REQUESTED** [review]

File: core/transformation/pathfinders/round_1/control_mappings.py (hunk 4 above)
Comment (MarcUsher) [reply core/transformation/pathfinders/round_1/control_mappings.py:124]: Ah no you're right, it's not checked in the same way by the general unit test as it's not a direct table mapping - writing a test now!

**Review by Joel-Pearce — APPROVED** [review]
LGTM

**Pull request merged**